From 3773eb0b426470c55f307c91d1506a4ff76e074a Mon Sep 17 00:00:00 2001
From: Steve Dickson <SteveD@redhat.com>
Date: Thu, 18 Sep 2014 14:20:17 -0400
Subject: [fs] nfs: nfs4_do_open should add negative results to the dcache

Message-id: <1411050057-12417-55-git-send-email-steved@redhat.com>
Patchwork-id: 96227
O-Subject: [RHEL7.1 PATCH 54/94] NFS: nfs4_do_open should add negative results to the dcache.
Bugzilla: 1111708
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>

From: NeilBrown <neilb@suse.de>

If you have an NFSv4 mounted directory which does not container 'foo'
and:

  ls -l foo
  ssh $server touch foo
  cat foo

then the 'cat' will fail (usually, depending a bit on the various
cache ages).  This is correct as negative looks are cached by default.
However with the same initial conditions:

  cat foo
  ssh $server touch foo
  cat foo

will usually succeed.  This is because an "open" does not add a
negative dentry to the dcache, while a "lookup" does.

This can have negative performance effects.  When "gcc" searches for
an include file, it will try to "open" the file in every director in
the search path.  Without caching of negative "open" results, this
generates much more traffic to the server than it should (or than
NFSv3 does).

The root of the problem is that _nfs4_open_and_get_state() will call
d_add_unique() on a positive result, but not on a negative result.
Compare with nfs_lookup() which calls d_materialise_unique on both
a positive result and on ENOENT.

This patch adds a call d_add() in the ENOENT case for
_nfs4_open_and_get_state() and also calls nfs_set_verifier().

With it, many fewer "open" requests for known-non-existent files are
sent to the server.

Signed-off-by: NeilBrown <neilb@suse.de>
Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Upstream Id: 4fa2c54b5198d09607a534e2fd436581064587ed
Bugzilla: 1111708
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfs/nfs4proc.c b/fs/nfs/nfs4proc.c
index 5785fe6..51d0fa7 100644
--- a/fs/nfs/nfs4proc.c
+++ b/fs/nfs/nfs4proc.c
@@ -2224,8 +2224,15 @@ static int _nfs4_open_and_get_state(struct nfs4_opendata *opendata,
  seq = raw_seqcount_begin(&sp->so_reclaim_seqcount);
 
  ret = _nfs4_proc_open(opendata);
- if (ret != 0)
+ if (ret != 0) {
+  if (ret == -ENOENT) {
+   d_drop(opendata->dentry);
+   d_add(opendata->dentry, NULL);
+   nfs_set_verifier(opendata->dentry,
+      nfs_save_change_attribute(opendata->dir->d_inode));
+  }
   goto out;
+ }
 
  state = nfs4_opendata_to_nfs4_state(opendata);
  ret = PTR_ERR(state);
-- 
1.7.1